README.md
![codecov](https://codecov.io/gh/CPrescher/xypattern/graph/badge.svg?token=05FUJFOV3R)
![CI](https://github.com/CPrescher/xypattern/actions/workflows/CI.yml/badge.svg)

# xypattern

## Description

A simple small library to handle x-y patterns, such as are collected with x-ray diffraction or Raman spectroscopy. 

## Installation

```bash
pip install xypattern
```

## Usage Examples

### Reading a file
```python
from xypattern import Pattern
import matplotlib.pyplot as plt

p1 = Pattern.from_file('path/to/file')
p1.scaling = 0.5
p1.offset = 0.1

plt.plot(p1.x, p1.y)
plt.show()
```

### Use a background pattern

```python

p2 = Pattern.from_file('path/to/file')
p2.scaling = 0.9
p1.background = p2

```

### Scale and stitch multiple patterns

```python
p1 = Pattern.from_file('path/to/file1')
p2 = Pattern.from_file('path/to/file2')
p3 = Pattern.from_file('path/to/file3')

from xypattern.combine import scale_patterns, stitch_patterns

patterns = [p1, p2, p3]
scale_patterns(patterns)
stitched_pattern = stitch_patterns(patterns)
``` 


CHANGELOG.md
# 1.0.5 (2013-11-14)

- enable poetry versioning plugin during build, to hopefully work when installing with pypi

# 1.0.4 (2023-11-13)

 - fix constraints to also actually work with python 3.8

# 1.0.3 (2023-11-13)

- provide packages for python 3.8 to 3.12

# 1.0.2 (2023-10-14)

## Bugfixes
- fix out of bound error for find_scaling (will be extrapolated in case there is a tiny mismatch in the binning)

# 1.0.1 (2023-10-14)

## new features
- find_scaling (and in accord also scale_patterns) works now with patterns that have different binning in the 
  overlapping region

# 1.0.0 (2023-10-13)

Initial Release

tests/test_util.py
import numpy as np
from xypattern import Pattern


def gaussian(x, mu, a, sig):
    return a * np.exp(-np.power(x - mu, 2.) / (2 * np.power(sig, 2.)))


def generate_peak_pattern(with_bkg=False):
    x = np.arange(0, 24, 0.01)
    y = np.zeros(x.shape)

    peaks = [
        [10, 3, 0.1],
        [12, 4, 0.1],
        [15, 6, 0.1],
    ]
    for peak in peaks:
        y += gaussian(x, peak[0], peak[1], peak[2])
    y_bkg = x * 0.4 + 5.0
    y_measurement = y + y_bkg

    if with_bkg:
        return Pattern(x, y_measurement), y_bkg
    return Pattern(x, y_measurement)


tests/test_signal.py
# -*- coding: utf-8 -*-
# Dioptas - GUI program for fast processing of 2D X-ray diffraction data
# Principal author: Clemens Prescher (clemens.prescher@gmail.com)
# Copyright (C) 2014-2019 GSECARS, University of Chicago, USA
# Copyright (C) 2015-2018 Institute for Geology and Mineralogy, University of Cologne, Germany
# Copyright (C) 2019-2020 DESY, Hamburg, Germany
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
import gc

from xypattern.util.signal import Signal, WeakRefList


def test_weakreflist_with_objects():
    weak_ref_list = WeakRefList()

    class A(str):
        pass

    a = A()
    weak_ref_list.append(a)
    assert len(weak_ref_list) == 1
    assert weak_ref_list[0]() == a

    del a
    gc.collect()
    assert len(weak_ref_list) == 0


def test_weakreflist_with_bound_methods():
    weak_ref_list = WeakRefList()

    class A:
        def method(self):
            return "lala"

    a = A()
    weak_ref_list.append(a.method)

    assert len(weak_ref_list) == 1
    assert weak_ref_list[0]()() == "lala"

    del a
    gc.collect()
    assert len(weak_ref_list) == 0


def test_weakreflist_with_insertion():
    weak_ref_list = WeakRefList()

    class A:
        def method(self):
            return "lala"

    a = A()
    weak_ref_list.insert(0, a.method)

    assert len(weak_ref_list) == 1
    assert weak_ref_list[0]()() == "lala"

    del a
    gc.collect()
    assert len(weak_ref_list) == 0


def test_signal_with_sole_functions():
    signal = Signal()
    memory = []

    def f():
        memory.append(1)

    signal.connect(f)
    signal.emit()
    assert memory == [1]

    signal.disconnect(f)
    signal.emit()
    assert memory == [1]


def test_signal_with_bound_methods():
    signal = Signal()
    memory = []

    class A:
        def method(self):
            memory.append(1)

    a = A()
    signal.connect(a.method)
    signal.emit()
    assert memory == [1]

    signal.disconnect(a.method)
    signal.emit()
    assert memory == [1]


def test_signal_with_signal():
    signal = Signal()
    memory = []

    signal2 = Signal()

    def f():
        memory.append(1)

    signal.connect(f)
    signal2.connect(signal)
    signal2.emit()

    assert memory == [1]


def test_signal_with_priority():
    signal = Signal()
    memory = []

    def f():
        memory.append(1)

    def g():
        memory.append(2)

    signal.connect(f)
    signal.connect(g, priority=True)
    signal.emit()

    assert memory == [2, 1]


tests/test_combine.py
import pytest

import numpy as np
from xypattern import Pattern
from xypattern.combine import find_overlap, find_scaling, scale_patterns, stitch_patterns


def test_find_overlap():
    p1 = Pattern([1, 2, 3], [1, 2, 3])
    p2 = Pattern([2, 3, 4], [2, 3, 4])
    assert find_overlap(p1, p2) == (2, 3)

    p3 = Pattern([4, 5, 6], [4, 5, 6])
    assert find_overlap(p1, p3) is None


def test_find_scaling():
    p1 = Pattern(np.array([1, 2, 3]), np.array([1, 1, 1]))
    p2 = Pattern(np.array([1, 2, 3]), np.array([2, 2, 2]))
    assert find_scaling(p1, p2) == 0.5

    p3 = Pattern(np.array([2, 3, 4]), np.array([4, 4, 2]))
    assert find_scaling(p1, p3) == 0.25


def test_find_scaling_with_different_x_binning():
    p1 = Pattern(np.array([1, 2, 3]), np.array([1, 1, 1]))
    p2 = Pattern(np.array([2, 2.5, 3, 3.5, 4]), np.array([2, 2, 2, 2, 2, ]))
    assert find_scaling(p1, p2) == 0.5
    assert find_scaling(p2, p1) == 2


def test_scale_patterns():
    p1 = Pattern(np.array([1, 2, 3]), np.array([1, 1, 1]))
    p2 = Pattern(np.array([1, 2, 3]), np.array([2, 2, 2]))
    p2.scaling = 3
    p3 = Pattern(np.array([2, 3, 4]), np.array([4, 4, 2]))

    scale_patterns([p1, p2, p3])
    assert p1.scaling == 1
    assert p2.scaling == 0.5
    assert p3.scaling == 0.25


def test_scale_patterns_unsorted():
    p1 = Pattern(np.array([1, 2, 3]), np.array([1, 1, 1]), name="p1")
    p2 = Pattern(np.array([5, 6, 7]), np.array([2, 2, 2]), name="p2")
    p3 = Pattern(np.array([3, 4, 5]), np.array([4, 4, 4]), name="p3")

    scale_patterns([p1, p2, p3])
    assert p1.scaling == 1
    assert p2.scaling == 0.5
    assert p3.scaling == 0.25


def test_scale_patterns_non_overlapping():
    p1 = Pattern(np.array([1, 2, 3]), np.array([1, 1, 1]), name="p1")
    p2 = Pattern(np.array([5, 6, 7]), np.array([2, 2, 2]), name="p2")

    with pytest.raises(ValueError):
        scale_patterns([p1, p2])


def generate_unsorted_patterns():
    p1 = Pattern(np.array([1, 2, 3]), np.array([1, 1, 1]), name="p1")
    p2 = Pattern(np.array([5, 6, 7]), np.array([2, 2, 2]), name="p2")
    p3 = Pattern(np.array([3, 4, 5, 6, 7, 8, 9]), np.array([4, 4, 4, 4, 4, 4, 4]), name="p3")
    p4 = Pattern(np.array([8, 9, 10, 11]), np.array([8, 8, 8, 8]), name="p4")
    return [p1, p2, p3, p4]


def test_scale_pattern_unsorted_different_length():
    [p1, p2, p3, p4] = generate_unsorted_patterns()

    scale_patterns([p1, p2, p3, p4])
    assert p1.scaling == 1
    assert p2.scaling == 0.5
    assert p3.scaling == 0.25
    assert p4.scaling == 0.125


def test_stitch_patterns():
    [p1, p2, p3, p4] = generate_unsorted_patterns()

    scale_patterns([p1, p2, p3, p4])
    p = stitch_patterns([p1, p2, p3, p4])
    assert p.x[0] == 1
    assert p.x[-1] == 11
    assert p.y[0] == 1
    assert np.array_equal(p.y, [1] * 11)
    assert p.scaling == 1
    assert p.x.shape == (11,)
    assert p.y.shape == (11,)
    assert p.x.shape == p.y.shape


xypattern/combine.py
from __future__ import annotations
import numpy as np
from scipy.interpolate import interp1d
from .pattern import Pattern


def stitch_patterns(patterns: list[Pattern], binning=None) -> Pattern:
    """
    Stitch together a list of patterns.
    :param patterns: list of patterns to be stitched together
    :param binning: binning to be applied to the stitched pattern, if None, the binning of the first pattern will be
                    used
    :return: stitched pattern
    """
    if binning is None:
        binning = patterns[0].x[1] - patterns[0].x[0]

    x = np.concatenate([pattern.x for pattern in patterns])
    y = np.concatenate([pattern.y for pattern in patterns])
    return Pattern(x, y).rebin(binning)


def scale_patterns(patterns: list[Pattern]):
    """
    Scales a list of patterns to the first pattern in respect to x.  The scaling will be in place setting the scale
    attribute of the patterns.
    :param patterns: list of patterns to be scaled
    """
    for pattern in patterns:
        pattern.scaling = 1

    sorted_patterns = sorted(patterns, key=lambda p: p.x[0])
    for ind, pattern in enumerate(sorted_patterns):
        if ind == 0:
            pattern.scaling = 1
            continue

        scale_ind = ind - 1
        scaling = find_scaling(sorted_patterns[scale_ind], pattern)
        while scaling is None:
            scale_ind -= 1
            if scale_ind < 0:
                raise ValueError("No overlap found between patterns")
            scaling = find_scaling(sorted_patterns[scale_ind], pattern)

        pattern.scaling = scaling


def find_overlap(p1: Pattern, p2: Pattern) -> tuple[float, float] | None:
    """
    Find the overlap in x between two patterns
    :param p1: pattern 1
    :param p2: pattern 2
    :return: tuple of x_min and x_max for the overlapping region or None if no overlap can be found
    """
    x_min = max(p1.x[0], p2.x[0])
    x_max = min(p1.x[-1], p2.x[-1])
    if x_min > x_max:
        return None
    return x_min, x_max


def find_scaling(p1: Pattern, p2: Pattern) -> float | None:
    """
    Find the scaling factor of p2 to p1
    :param p1: pattern 1
    :param p2: pattern 2
    :return: scaling factor
    """
    overlap = find_overlap(p1, p2)
    if overlap is None:
        return None

    p1_indices = np.where((p1.x >= overlap[0]) & (p1.x <= overlap[1]))
    p2_indices = np.where((p2.x >= overlap[0]) & (p2.x <= overlap[1]))
    x1 = p1.x[p1_indices]
    x2 = p2.x[p2_indices]
    y1 = p1.y[p1_indices]
    y2 = p2.y[p2_indices]

    if len(x1) == len(x2) and np.allclose(x1, x2):
        return np.mean(y1 / y2)

    f2 = interp1d(x2, y2, kind="linear", fill_value="extrapolate")
    p2_interpolated = f2(x1)
    return np.mean(y1 / p2_interpolated)


xypattern/util/signal.py
# -*- coding: utf-8 -*-
# Dioptas - GUI program for fast processing of 2D X-ray diffraction data
# Principal author: Clemens Prescher (clemens.prescher@gmail.com)
# Copyright (C) 2014-2019 GSECARS, University of Chicago, USA
# Copyright (C) 2015-2018 Institute for Geology and Mineralogy, University of Cologne, Germany
# Copyright (C) 2019-2020 DESY, Hamburg, Germany
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import inspect
import weakref

__export__ = ["Signal"]


class Signal:
    def __init__(self, *_):
        self.listeners = WeakRefList()
        self.priority_listeners = WeakRefList()
        self.blocked = False

    def connect(self, handle, priority=False):
        """
        Connects a function handle to the Signal.
        :param handle: function handle to be called when the signal is emitted
        :param priority: if set to True, the handle will be added as the first function to be called in the case of
                        the event
        If multiple handles are added with priority, they will obviously be called in the reverse order of adding.
        """
        if priority:
            self.priority_listeners.insert(0, handle)
        else:
            self.listeners.append(handle)

    def disconnect(self, handle):
        """
        Removes a certain function handle from the list of listeners to be called in case of Signal emitted.
        :param handle: function handle to be removed from the listeners
        """
        try:
            self.listeners.remove(handle)
        except ValueError:
            pass

        try:
            self.priority_listeners.remove(handle)
        except ValueError:
            pass

    def emit(self, *args):
        if self.blocked:
            return
        self._serve_listeners(self.priority_listeners, *args)
        self._serve_listeners(self.listeners, *args)

    @staticmethod
    def _serve_listeners(listeners, *args):
        for ref in listeners:
            handle = ref()
            if type(handle) == Signal:
                handle.emit(*args)
            else:
                try:
                    handle(*args)
                except AttributeError:
                    handle()

    def clear(self):
        """
        Removes all listeners from the Signal.
        """
        self.listeners = WeakRefList()
        self.priority_listeners = WeakRefList()


class WeakRefList(list):
    """
    A list which holds weak references to its items. If an item is deleted, the reference to it will be removed from
    the list. This is useful for Signals, where we want to hold a list of listeners, but don't want to prevent the
    garbage collector from deleting the listeners.
    It is not a full reimplementation, only the methods which are used in the Signal class are implemented - append,
    remove, insert. This list will work for object methods as well as objects. To retrieve the orginal item, the
    value of the weak reference has to be called. E.g.:
    >>> class A:
    >>>     def method(self):
    >>>         return "lala"
    >>>
    >>> a = A()
    >>> weak_ref_list = WeakRefList()
    >>> weak_ref_list.append(a.method)
    >>> weak_ref_list[0]()() == "lala"
    """
    def append(self, item):
        super(WeakRefList, self).append(self._ref(item))

    def remove(self, item):
        super(WeakRefList, self).remove(self._ref(item))

    def insert(self, index, item):
        super(WeakRefList, self).insert(index, self._ref(item))

    def _remove_ref(self, ref):
        super(WeakRefList, self).remove(ref)

    def _ref(self, item):
        if inspect.ismethod(item):
            return weakref.WeakMethod(item, self._remove_ref)
        else:
            return weakref.ref(item, self._remove_ref)


xypattern/scripts/stitch_patterns.py
import argparse
from .. import Pattern
from ..combine import stitch_patterns, scale_patterns


def main():
    parser = argparse.ArgumentParser(description='Stitch patterns together',
                                     prog='stitch_patterns.py')
    parser.add_argument('patterns', metavar='pattern', type=str, nargs='+',
                        help='patterns to be stitched together')
    parser.add_argument('-o', '--output', metavar='output', type=str, nargs=1,
                        help='output file name')
    parser.add_argument('-b', '--binning', metavar='binning', type=float, nargs=1,
                        help='binning to be applied to the stitched pattern, if None, the binning of the first pattern '
                             'will be used')
    parser.add_argument('-s', '--scale', action='store_true',
                        help='scale patterns to the first pattern in respect to x')
    args = parser.parse_args()

    patterns = [Pattern.from_file(pattern) for pattern in args.patterns]
    if args.scale:
        scale_patterns(patterns)

    pattern = stitch_patterns(patterns, binning=args.binning[0] if args.binning else None)
    pattern.save(args.output[0] if args.output else 'stitched.xy')


xypattern/pattern.py
# -*- coding: utf-8 -*-
from __future__ import annotations
import os

import numpy as np
from scipy.interpolate import interp1d
from scipy.ndimage import gaussian_filter1d

from .util.signal import Signal
from .auto_background import AutoBackground, SmoothBrucknerBackground


class Pattern(object):
    """
    A Pattern is a set of x and y values.
    It can be loaded from a file or created from scratch and can be modified by
    different methods.
    It builds the basis for all calculations in glassure.

    :param x: x values of the pattern
    :param y: y values of the pattern
    :param name: name of the pattern
    """

    def __init__(self, x: np.ndarray = None, y: np.ndarray = None, name: str = ''):
        """
        Creates a new Pattern object, x and y should have the same shape.
        """
        if x is None:
            self._original_x = np.linspace(0.1, 15, 100)
        else:
            self._original_x = x
        if y is None:
            self._original_y = np.log(self._original_x ** 2) - (self._original_x * 0.2) ** 2
        else:
            self._original_y = y

        self.name = name
        self._offset = 0.0
        self._scaling = 1.0
        self._smoothing = 0.0
        self._background_pattern = None

        self._auto_bkg: AutoBackground = None

        self._pattern_x = self._original_x
        self._pattern_y = self._original_y

        self._auto_background_before_subtraction_pattern = None
        self._auto_background_pattern = None

        self.changed = Signal()

    def load(self, filename: str, skiprows: int = 0):
        """
        Loads a pattern from a file. The file can be either a .xy or a .chi file. The .chi file will be loaded with
        skiprows=4 by default.

        :param filename: path to the file
        :param skiprows: number of rows to skip when loading the data (header)
        """
        try:
            if filename.endswith('.chi'):
                skiprows = 4
            data = np.loadtxt(filename, skiprows=skiprows)
            self._original_x = data.T[0]
            self._original_y = data.T[1]
            self.name = os.path.basename(filename).split('.')[:-1][0]
            self.recalculate_pattern()

        except ValueError:
            raise ValueError('Wrong data format for pattern file! - ' + filename)

    @staticmethod
    def from_file(filename: str, skip_rows: int = 0) -> Pattern | '-1':
        """
        Loads a pattern from a file. The file can be either a .xy or a .chi file. The .chi file will be loaded with
        skiprows=4 by default.

        :param filename: path to the file
        :param skip_rows: number of rows to skip when loading the data (header)
        """
        try:
            if filename.endswith('.chi'):
                skip_rows = 4
            data = np.loadtxt(filename, skiprows=skip_rows)
            x = data.T[0]
            y = data.T[1]
            name = os.path.basename(filename).split('.')[:-1][0]
            return Pattern(x, y, name)

        except ValueError:
            raise ValueError('Wrong data format for pattern file! - ' + filename)

    def save(self, filename, header='', subtract_background=False, unit='2th_deg'):
        """
        Saves the x, y data to file. Supporting several file formats: .chi, .xy, .fxye
        :param filename: where to save the data
        :param header: header for file
        :param subtract_background: whether to save subtracted data
        :param unit: x-unit used for the standard chi header (unused for other formats)
        """
        x, y = self.data
        if not subtract_background and self.background_pattern is not None:
            y += self.background_pattern.y

        num_points = len(x)

        file_handle = open(filename, 'w')

        if filename.endswith('.chi'):
            if header is None or header == '':
                file_handle.write(filename + '\n')
                file_handle.write(unit + '\n\n')
                file_handle.write("       {0}\n".format(num_points))
            else:
                file_handle.write(header)
            for ind in range(num_points):
                file_handle.write(' {0:.7E}  {1:.7E}\n'.format(x[ind], y[ind]))
        elif filename.endswith('.fxye'):
            factor = 100
            if 'CONQ' in header:
                factor = 1
            header = header.replace('NUM_POINTS', '{0:.6g}'.format(num_points))
            header = header.replace('MIN_X_VAL', '{0:.6g}'.format(factor * x[0]))
            header = header.replace('STEP_X_VAL', '{0:.6g}'.format(factor * (x[1] - x[0])))

            file_handle.write(header)
            file_handle.write('\n')
            for ind in range(num_points):
                file_handle.write('\t{0:.6g}\t{1:.6g}\t{2:.6g}\n'.format(factor * x[ind], y[ind], np.sqrt(abs(y[ind]))))
        else:

            data = np.dstack((x, y))
            np.savetxt(file_handle, data[0], header=header)
        file_handle.close()

    @property
    def background_pattern(self) -> Pattern:
        """
        Returns the background pattern of the current pattern.
        :return: background Pattern
        """
        return self._background_pattern

    @background_pattern.setter
    def background_pattern(self, pattern: Pattern | None):
        if self._background_pattern is not None:
            self._background_pattern.changed.disconnect(self.recalculate_pattern)

        self._background_pattern = pattern
        if self._background_pattern is not None:
            self._background_pattern.changed.connect(self.recalculate_pattern)
        self.recalculate_pattern()

    def rebin(self, bin_size: float) -> Pattern:
        """
        Returns a new pattern, which is a rebinned version of the current one.

        :param bin_size: Size of the bins
        :return: rebinned Pattern
        """
        x, y = self.data
        x_min = np.round(np.min(x) / bin_size) * bin_size
        x_max = np.round(np.max(x) / bin_size) * bin_size
        new_x = np.arange(x_min, x_max + 0.1 * bin_size, bin_size)

        bins = np.hstack((x_min - bin_size * 0.5, new_x + bin_size * 0.5))
        new_y = (np.histogram(x, bins, weights=y)
                 [0] / np.histogram(x, bins)[0])

        return Pattern(new_x, new_y)

    @property
    def data(self) -> tuple[np.ndarray, np.ndarray]:
        """
        Returns the data of the pattern. If a background pattern is set, the background will be subtracted from the
        pattern. If smoothing is set, the pattern will be smoothed.

        :return: Tuple of x and y values
        """
        return self._pattern_x, self._pattern_y

    def recalculate_pattern(self):
        """
        Returns the data of the pattern. If a background pattern is set, the background will be subtracted from the
        pattern. If smoothing is set, the pattern will be smoothed.

        :return: Tuple of x and y values
        """
        if self._background_pattern is not None:
            # create background function
            x_bkg, y_bkg = self._background_pattern.data

            if not np.array_equal(x_bkg, self._original_x):
                # the background will be interpolated
                f_bkg = interp1d(x_bkg, y_bkg, kind='linear')

                # find overlapping x and y values:
                ind = np.where((self._original_x <= np.max(x_bkg)) &
                               (self._original_x >= np.min(x_bkg)))
                x = self._original_x[ind]
                y = self._original_y[ind]

                if len(x) == 0:
                    # if there is no overlapping between background and pattern, raise an error
                    raise BkgNotInRangeError(self.name)

                y = y * self._scaling + self.offset - f_bkg(x)
            else:
                # if pattern and bkg have the same x basis we just delete y-y_bkg
                x, y = self._original_x, self._original_y * self._scaling + self.offset - y_bkg
        else:
            x, y = self.original_data
            y = y * self.scaling + self.offset

        if self._auto_bkg is not None:
            self._auto_background_before_subtraction_pattern = Pattern(x, y)
            x, y = self._auto_background_before_subtraction_pattern.limit(*self._auto_bkg_roi).data
            y_bkg = self._auto_bkg.extract_background(Pattern(x, y))
            self._auto_background_pattern = Pattern(x, y_bkg, name='auto_bkg_' + self.name)
            y -= y_bkg

        if self.smoothing > 0:
            y = gaussian_filter1d(y, self.smoothing)

        self._pattern_x = x
        self._pattern_y = y
        self.changed.emit()

    @data.setter
    def data(self, data: tuple[np.ndarray, np.ndarray]):
        """
        Sets the data of the pattern. Also resets the scaling and offset to 1 and 0 respectively.

        :param data: tuple of x and y values
        """
        (x, y) = data
        self._original_x = x
        self._original_y = y
        self.scaling = 1.0
        self.offset = 0

    @property
    def original_data(self) -> tuple[np.ndarray, np.ndarray]:
        """
        Returns the original data of the pattern without any background subtraction or smoothing.

        :return: tuple of x and y values
        """
        return self._original_x, self._original_y

    @property
    def x(self) -> np.ndarray:
        """ Returns the x values of the pattern """
        return self._pattern_x

    @x.setter
    def x(self, new_value: np.ndarray):
        """ Sets the x values of the pattern """
        self._original_x = new_value
        self.recalculate_pattern()

    @property
    def y(self) -> np.ndarray:
        """ Returns the y values of the pattern """
        return self._pattern_y

    @y.setter
    def y(self, new_y: np.ndarray):
        """ Sets the y values of the pattern """
        self._original_y = new_y
        self.recalculate_pattern()

    @property
    def scaling(self) -> float:
        """ Returns the scaling of the pattern """
        return self._scaling

    @scaling.setter
    def scaling(self, value):
        """ 
        Sets the scaling of the pattern, if below 0, it will be set to 0 
        instead.
        """
        if value < 0:
            self._scaling = 0.0
        else:
            self._scaling = value
        self.recalculate_pattern()

    @property
    def offset(self) -> float:
        """ Returns the offset of the pattern """
        return self._offset

    @offset.setter
    def offset(self, value):
        """ Sets the offset of the pattern """
        self._offset = value
        self.recalculate_pattern()

    @property
    def smoothing(self) -> float:
        """ Returns the smoothing of the pattern """
        return self._smoothing

    @smoothing.setter
    def smoothing(self, value):
        """ Sets the smoothing of the pattern """
        self._smoothing = value
        self.recalculate_pattern()

    def set_auto_background_subtraction(self, parameters: list[float], roi: list[float] = None, recalc_pattern=True):
        """
        Sets an automatic background subtraction to the pattern. The background will be subtracted from the pattern
        when calling the data property.

        :param parameters: list of parameters for the automatic background subtraction. The first parameter is the
        smoothing amount, the second, the number of iterations for the bruckner smoothing and the third the order of the
        chebyshev polynomial.
        :param roi: region of interest to be used for  the background subtraction. If None, the whole pattern will be
        used.
        """
        if roi is None:
            min_step = self.x[1] - self.x[0]
            max_step = self.x[-1] - self.x[-2]
            roi = [np.min(self.x) - min_step, np.max(self.x) + max_step]
        self._auto_bkg = SmoothBrucknerBackground(*parameters)
        self._auto_bkg_roi = roi
        if recalc_pattern:
            self.recalculate_pattern()

    def unset_auto_background_subtraction(self):
        self._auto_bkg = None
        self.recalculate_pattern()

    @property
    def auto_background_pattern(self) -> Pattern:
        """
        Returns the auto background pattern
        :return: background Pattern
        """
        return self._auto_background_pattern

    @property
    def auto_background_before_subtraction_pattern(self) -> Pattern:
        """
        Returns the pattern before the auto background subtraction
        :return: background Pattern
        """
        return self._auto_background_before_subtraction_pattern

    def limit(self, x_min: float, x_max: float) -> Pattern:
        """
        Limits the pattern to a specific x-range. Does not modify inplace but returns a new limited Pattern

        :param x_min: lower limit of the x-range
        :param x_max: upper limit of the x-range
        :return: limited Pattern
        """
        x, y = self.data
        return Pattern(x[np.where((x_min < x) & (x < x_max))],
                       y[np.where((x_min < x) & (x < x_max))])

    def extend_to(self, x_value: float, y_value: float) -> Pattern:
        """
        Extends the current pattern to a specific x_value by filling it with the y_value. Does not modify inplace but
        returns a new filled Pattern

        :param x_value: Point to which extending the pattern should be smaller than the lowest x-value in the pattern or
        vice versa
        :param y_value: number to fill the pattern with
        :return: extended Pattern
        """
        x_step = np.mean(np.diff(self.x))
        x_min = np.min(self.x)
        x_max = np.max(self.x)
        if x_value < x_min:
            x_fill = np.arange(x_min - x_step, x_value -
                               x_step * 0.5, -x_step)[::-1]
            y_fill = np.zeros(x_fill.shape)
            y_fill.fill(y_value)

            new_x = np.concatenate((x_fill, self.x))
            new_y = np.concatenate((y_fill, self.y))
        elif x_value > x_max:
            x_fill = np.arange(x_max + x_step, x_value + x_step * 0.5, x_step)
            y_fill = np.zeros(x_fill.shape)
            y_fill.fill(y_value)

            new_x = np.concatenate((self.x, x_fill))
            new_y = np.concatenate((self.y, y_fill))
        else:
            return self

        return Pattern(new_x, new_y)

    def to_dict(self) -> dict:
        """
        Returns a dictionary representation of the pattern which can be used to save the pattern to a json file.

        :return: dictionary representation of the pattern
        """
        return {
            'name': self.name,
            'x': self._original_x.tolist(),
            'y': self._original_y.tolist(),
            'scaling': self.scaling,
            'offset': self.offset,
            'smoothing': self.smoothing,
            'bkg_pattern': self._background_pattern.to_dict() if self._background_pattern is
                                                                 not None else None
        }

    @staticmethod
    def from_dict(json_dict: dict) -> Pattern:
        """
        Creates a new Pattern from a dictionary representation of a Pattern.

        :param json_dict: dictionary representation of a Pattern
        :return: new Pattern
        """
        pattern = Pattern(np.array(json_dict['x']),
                          np.array(json_dict['y']),
                          json_dict['name'])

        pattern.scaling = json_dict['scaling']
        pattern.offset = json_dict['offset']

        if json_dict['bkg_pattern'] is not None:
            bkg_pattern = Pattern.from_dict(json_dict['bkg_pattern'])
        else:
            bkg_pattern = None
        pattern.background_pattern = bkg_pattern

        pattern.smoothing = json_dict['smoothing']
        pattern.recalculate_pattern()

        return pattern

    ###########################################################
    # Operators:

    def __sub__(self, other: Pattern) -> Pattern:
        """
        Subtracts the other pattern from the current one. If the other pattern
        has a different shape, the subtraction will be done on the overlapping
        x-values and the background will be interpolated. If there is no
        overlapping between the two patterns, a BkgNotInRangeError will be
        raised.

        :param other: Pattern to be subtracted
        :return: new Pattern
        """
        orig_x, orig_y = self.data
        other_x, other_y = other.data

        if orig_x.shape != other_x.shape:
            # the background will be interpolated
            other_fcn = interp1d(other_x, other_y, kind='cubic')

            # find overlapping x and y values:
            ind = np.where((orig_x <= np.max(other_x)) & (orig_x >= np.min(other_x)))
            x = orig_x[ind]
            y = orig_y[ind]

            if len(x) == 0:
                # if there is no overlapping between background and pattern, raise an error
                raise BkgNotInRangeError(self.name)
            return Pattern(x, y - other_fcn(x))
        else:
            return Pattern(orig_x, orig_y - other_y)

    def __add__(self, other: Pattern) -> Pattern:
        """
        Adds the other pattern to the current one. If the other pattern
        has a different shape, the addition will be done on the overlapping
        x-values and the y-values of the other pattern will be interpolated. 
        If there is no overlapping between the two patterns, a BkgNotInRangeror 
        will be raised.

        :param other: Pattern to be added
        :return: new Pattern
        """
        orig_x, orig_y = self.data
        other_x, other_y = other.data

        if orig_x.shape != other_x.shape:
            # the background will be interpolated
            other_fcn = interp1d(other_x, other_y, kind='linear')

            # find overlapping x and y values:
            ind = np.where((orig_x <= np.max(other_x)) & (orig_x >= np.min(other_x)))
            x = orig_x[ind]
            y = orig_y[ind]

            if len(x) == 0:
                # if there is no overlapping between background and pattern, raise an error
                raise BkgNotInRangeError(self.name)
            return Pattern(x, y + other_fcn(x))
        else:
            return Pattern(orig_x, orig_y + other_y)

    def __rmul__(self, other: float) -> Pattern:
        """
        Multiplies the pattern with a scalar.

        :param other: scalar to multiply with
        :return: new Pattern
        """
        orig_x, orig_y = self.data
        return Pattern(np.copy(orig_x), np.copy(orig_y) * other)

    def __eq__(self, other: Pattern) -> bool:
        """
        Checks if two patterns are equal. Two patterns are equal if their data
        is equal.

        :param other: Pattern to compare with
        :return: True if equal, False otherwise
        """
        if not isinstance(other, Pattern):
            return False
        if np.array_equal(self.data, other.data):
            return True
        return False

    def __len__(self):
        return len(self.x)

    def __str__(self):
        return f'Pattern \'{self.name}\' with {len(self)} points'


class BkgNotInRangeError(Exception):
    def __init__(self, pattern_name: str):
        self.pattern_name = pattern_name

    def __str__(self):
        return "The background range does not overlap with the Pattern range for " + self.pattern_name
